src/Bicep.Core/Semantics/ParamsSemanticModel.cs

```diff
@@ -67,24 +140,40 @@ private IEnumerable<IDiagnostic> GetAdditionalSemanticDiagnostics()
         }
         
-        private void GetParameterMismatchDiagnostics(IDiagnosticWriter diagnosticWriter, ImmutableArray<Metadata.ParameterMetadata> parameters, IEnumerable<ParameterAssignmentSyntax> parameterAssignments)
+        private void GetParameterMismatchDiagnostics(IDiagnosticWriter diagnosticWriter, IEnumerable<ParameterDeclarationSyntax> parameters, IEnumerable<ParameterAssignmentSyntax> parameterAssignments)
         {
-            var parametersDict = parameters.ToDictionary(x => x.Name, LanguageConstants.IdentifierComparer);
-            var parametersAssignmentsDict = parameterAssignments.ToDictionary(x => x.Name.IdentifierName, LanguageConstants.IdentifierComparer);
-            
-            // parameters that are declared but not assigned
-            var missingRequiredParams = parametersDict.Values
-                .Where(x => x.IsRequired)
-                .Where(x => !parametersAssignmentsDict.ContainsKey(x.Name))
-                .OrderBy(x => x.Name)
-                .Select(x => x.Name)
-                .ToArray();
             // parameters that are assigned but not declared
-            var missingAssignedParams = parametersAssignmentsDict
-                .Where(x => !parametersDict.ContainsKey(x.Key))
-                .Select(x => x.Value)
-                .OrderBy(x => x.Name.IdentifierName);
+            var missingAssignedParams = new List<ParameterAssignmentSyntax>();
+
+            foreach (var parameterAssignment in parameterAssignments)
+            {
+                if (this.ParamBinder.GetSymbolInfo(parameterAssignment) is ParameterAssignmentSymbol symbol && GetParameterDeclaration(symbol) is null)
+                {
+                    missingAssignedParams.Add(parameterAssignment);
+                }
+            }
+
+            // parameters that are declared but not assigned
+            var missingRequiredParams = new List<String>();
+
+            if (this.BicepCompilation is null)
+            {
+                return;
+            }
+
+            var parameterMetadata = this.BicepCompilation.GetEntrypointSemanticModel().Parameters;
+            var metadataByName = parameterMetadata.ToDictionary(x => x.Name, x => x);
 
+            foreach (var parameter in parameters)
+            {
+                if (this.BicepCompilation.GetEntrypointSemanticModel().Binder.GetSymbolInfo(parameter) is ParameterSymbol symbol && GetParameterAssignment(symbol) is null &&
+                    metadataByName[parameter.Name.IdentifierName].IsRequired)
+                {
+                    missingRequiredParams.Add(parameter.Name.IdentifierName);
+                }
+            }
+            
+            // emit diagnostic only if there is a using statement
             var usingDeclarationSyntax = BicepParamFile.ProgramSyntax.Children.OfType<UsingDeclarationSyntax>().SingleOrDefault();
             if (usingDeclarationSyntax is not null && missingRequiredParams.Count() > 0)
```

Comment:
This is related to [my other comment ](https://github.com/Azure/bicep/pull/7823/files#r942742463) - `missingRequiredParams.Count()` can be replaced with `missingRequiredParams.Any()`.